### DISCUSSION TITLE
Add `origin` parameter to `send` method of an Interpreter

### DISCUSSION THREAD

**Discussion by ycaptain** [statelyai/xstate#2014] [Ideas]
I wang to empower a commen event more like an SCXML event.

There is no global context to manage the namespace of machines. So we cannot send an event to another machine easily.
But we can manage the global namespace ourselves and dynamically create machines with our context.
Then ref to other machines could be reached in runtime and call `send` method of other machines in our actions or whatever.

Although we could call `send` directly with payload, there is no information about the event's origin (The inspector cannot generate a sequence step by calling the machine's `send` method directly even in an action).

The point is there is only one way to implement this currently if I want the event to have an origin machine (so the inspector could generate a sequence diagram well): **constructing an SCXML event to the `send` method**.

Write an SCXML event is very inconvenient.

Ex, if I want to construct an SCXML event to represent an event whose type is `foo`:
```js
// SCXML
anotherMachine.send({
    $$type: 'scxml',
    type: 'external',
    data: {
        type: 'foo',
        ...payload,
    },
    name: 'foo',
    origin: myMachineSessionId
});

// no origin send
anotherMachine.send({
    type: 'foo',
}, payload);
```

Maybe we can add the `origin` parameter to `send` method of `Interpreter`. Like:
```js
anotherMachine.send({
    type: 'foo',
    origin: myMachineSessionId,
}, payload);
```

BTW, I am not sure if it's a good practice to call other machines' `send` method in actions.

Insight from https://github.com/davidkpiano/xstate/discussions/1773#discussioncomment-238565

**Comment by ycaptain** [comment]
I noticed the helper function `toSCXMLEvent` from [the service example of @xstate/inspect](https://github.com/davidkpiano/xstate/blob/master/packages/xstate-inspect/examples/server.ts). It could help me to construct an SCXML event.

**Reply by davidkpiano** [reply]
Yes, you can do this:

```js
someService.send(toSCXMLEvent({ type: 'foo' }, { origin: ... })
```

**Comment by ycaptain** [comment]
`toSCXML` makes this easier.

However, I still wonder about the roles of a common event and an SCXML event in `xstate`. 

Is an SCXML event more powerful than a common event?
Do we limit the usage of a common event for some reasons?
Do you guys think it is acceptable to construct an SCXML event to enhance the power of the `send` method?
Should we make a common event as same an SCXML event?

I want to listen to people's ideas and suggestions, and try to post a PR for adding the `origin` parameter to the `send` method.

**Comment by seloner** [comment]
I also think this is a good change .

**Reply by ycaptain** [reply]
Thanks for your feedback.👏
Can you give an example about when you need this feature?
